### ISSUE TITLE
[Dynamo] Cache miss due to strict guards for int/float variables

### ISSUE THREAD

**Issue by zyq8709** [pytorch/pytorch#93344]
### 🚀 The feature, motivation and pitch

For this code:
```
@dynamo.optimize("eager")
def test(boxes: torch.Tensor, mix: Tuple[int, int]):
    x1, y1, x2, y2 = boxes[:, 0], boxes[:, 1], boxes[:, 2], boxes[:, 3]
    w = (x2 - x1 + 1)
    h = (y2 - y1 + 1)
    new = mix[0] * mix[1]
    cx = x1 + 0.5 * w
    cy = y1 + 0.5 * h
    return cx, cy, w, h, new

#warmup
boxes = torch.randn(1000, 4, device="cuda:0")
mix = tuple(torch.randint(30, (2,)).tolist())
test(boxes, mix)

for i in range(10):
    boxes = torch.randn(1000, 4, device="cuda:0")
    mix = tuple(torch.randint(30, (2,)).tolist())
    test(boxes, mix)
```
When running the code, cache missing will occur in the 10 iterations because the data changes each time. For “mix”, the generated guard is as followed:
```
local 'mix' EQUALS_MATCH
{
    'guard_types': ['LIST_LENGTH', 'EQUALS_MATCH'],
    'code': ['___check_type_id(mix, 94705134607264)', 'len(mix) == 2', '___check_type_id(mix[0], 94705134621472)', '___check_type_id(mix[1], 94705134621472)', 'mix == (6, 24)'],
} 
```
For each of its elements, guards are:
```
local 'mix[0]' CONSTANT_MATCH
{
    'guard_types': ['EQUALS_MATCH'],
    'code': ['___check_type_id(mix[0], 94705134621472)', 'mix[0] == 6'],
}

local 'mix[1]' CONSTANT_MATCH
{
    'guard_types': ['EQUALS_MATCH'],
    'code': ['___check_type_id(mix[1], 94705134621472)', 'mix[1] == 24'],
}
```
These all come from the code in “VariableBuilder._wrap” :
```
def _wrap(self, value):
        ......
        elif istype(value, (tuple, list, odict_values)) or is_namedtuple(value):
            # One can index a tensor with a list/tuple. Therefore, we need to
            # have a stricter match.
            if istype(value, (tuple, list)) and all(
                [isinstance(x, int) or is_numpy_int_type(x) or x is None for x in value]
            ):
                guards = self.make_guards(GuardBuilder.EQUALS_MATCH)
            else:
                guards = self.make_guards(GuardBuilder.LIST_LENGTH)
            output = [
                VariableBuilder(self.tx, GetItemSource(self.get_source(), i))(
                    item
                ).add_guards(guards)
                for i, item in enumerate(value)
            ]
            result = self.list_type(value)(output, guards=guards)
            if istype(value, list):
                return self.tx.output.side_effects.track_list(
                    self.source, value, result
                )
            return result
        ......
        elif ConstantVariable.is_literal(value) or istype(
            value, (torch.Size, torch.device, torch.dtype)
        ):
            if type(value) in (int, float) and not config.specialize_int_float:
                # unspecializing int/float by default, but still
                # specialize for the following conditions
                if (
                    value in self._common_constants()
                    or isinstance(self.source, GlobalSource)
                    or isinstance(self.source, GetItemSource)
                    or (
                        isinstance(self.source, AttrSource)
                        and isinstance(self.source.base, GlobalSource)
                    )
                ):
                    return ConstantVariable(
                        value=value,
                        guards=make_guards(GuardBuilder.CONSTANT_MATCH),
                    )
                else:
                    return self.wrap_unspecialized_primitive(value)
            else:
                return ConstantVariable(
                    value=value,
                    guards=make_guards(GuardBuilder.CONSTANT_MATCH),
                )
        ......
```
Therefore, for int/float variables and their related complex data structures,  value equality is checked, which is more strict than tensors (only check shape, dim, requires_grad, dtype etc, no value check). Of course, this would help constant folding and reduce ops in generated fx Graph. 
Take "add" BuiltinVariable for example:
```
def call_function(
        self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
    ) -> "VariableTracker":
        ......
        handler = getattr(self, f"call_{self.fn.__name__}", None)
        ......
        if handler:
            try:
                result = handler(tx, *args, **kwargs)
        ......

def call_add(self, tx, *args, **kwargs):
        return args[0].call_method(tx, "__add__", args[1:], kwargs)
```
This leads to ConstantVariable.call_method:
```
def call_method(
        self,
        tx,
        name,
        args: "List[VariableTracker]",
        kwargs: "Dict[str, VariableTracker]",
    ) -> "VariableTracker":
       
        ......
        elif has_arith_binop(int) or has_arith_binop(float):
            op = getattr(operator, name)
            add_target = const_args[0]
            ......
            return ConstantVariable(op(self.value, add_target), **options)
        ......
``````
As a result, the "add" operation will be completed during InstructionTranslator.step and no node will be inserted in fx Graph. This indeed relies on EQUALS_MATCH and incurs the cache missing when data value changes. 

Through testing, bad performance can be observed, even worse than purely eager execution due to caching missing and recompiling each time.

### Alternatives

For int/float relevant variables, perhaps we can provide a switch in config to reduce cache missing. 
When the switch is on, these variables are not wrapped in ConstantVariable and not related to EQUALS_MATCH guards. Instead, we can wrap them into DynamicShapeVariable, however along with many code changes in different files mainly in "variable" directory and guards.py.

**Comment by yanboliang** [comment]
We had a config to switch specializing vs unspecializing python int/float, it wraps python int/float as 1-element tensor and guards on ```TYPE_MATCH```. 
https://github.com/pytorch/pytorch/blob/438f12d91a8adec94145b51181db310dccb4e652/torch/_dynamo/config.py#L46

**Comment by yanboliang** [comment]
I think we can't relax the constraint if the int/float is part of a list, we assume the list of int/float should be constant so far, but we can re-evaluate this decision.

**Comment by ezyang** [comment]
This is dynamic shapes related, sort of, because unspecialized ints can also be treated as SymInt so we don't have to model it as a 1 elem tensor which is buggy. See also https://github.com/pytorch/pytorch/pull/93150

**Comment by zyq8709** [comment]
For the above demo, when specialize_int_float is set to False, cache missing still happens, because mix[0]'s and mix[1]'s sources are GetItemSource, which meets the following conditions and still leads to specialization. To be honest, I don't quite understand the logic behind these source related conditions.
```
            if type(value) in (int, float) and not config.specialize_int_float:
                # unspecializing int/float by default, but still
                # specialize for the following conditions
                if (
                    value in self._common_constants()
                    or isinstance(self.source, GlobalSource)
                    or isinstance(self.source, GetItemSource)
                    or (
                        isinstance(self.source, AttrSource)
                        and isinstance(self.source.base, GlobalSource)
                    )
                ):
                    return ConstantVariable(
                        value=value,
                        guards=make_guards(GuardBuilder.CONSTANT_MATCH),
                    )
```

**Comment by yanboliang** [comment]
> For the above demo, when specialize_int_float is set to False, cache missing still happens. One reason is that the guard for tuple is the same as LIST_MATCH and EQUALS_MATCH. The other reason is that mix[0]'s and mix[1]'s sources are GetItemSource, which meets the following conditions and still leads to specialization. To be honest, I don't quite understand the logic behind these source related conditions.

@zyq8709 Yes, if you pass in a list of int/floats, we don't support unspecialize them. I think this decision was made based on the data from benchmarks(TB/HF/TIMM). In most of the cases, users pass in a list of int/floats, they expect they are constants. To mitigate this, you can pass directly as arguments rather than a list and set ```specialize_int_float=False```, it should work well.

**Comment by yanboliang** [comment]
> The conditions need deep cleaning. If it doesn't make sense, it's probably because it doesn't make sense.

Agreed, I think we should clean up these conditions after we have SymInt/SymFloat as #93150

**Comment by zyq8709** [comment]
As for the guard of mix,
```
['___check_type_id(mix, 94705134607264)', 'len(mix) == 2', '___check_type_id(mix[0], 94705134621472)', '___check_type_id(mix[1], 94705134621472)', 'mix == (6, 24)']
``` 
I think maybe `mix == (6, 24)` is not necessary for EQUALS_MATCH of tuple/list. With or without int/float specialized, I feel removing this part is fine. 
Could you tell me the consideration on this part or some counter-examples which lead to incorrectness without that part?
